File: ch2/huffmantree.cpp
/*
 * huffmantree.cpp
 *
 *      Author: flamingo
 *      E-mail: epiforce57@gmail.com
 */

#include "huffmantree.h"

#include <algorithm>
#include <exception>
#include <iostream>

bool NodeCmpFunc(const HuffmannTree::Node *n1, const HuffmannTree::Node *n2)
{
    return n1->weight < n2->weight;
}

HuffmannTree::Node * HuffmannTree::createTree(vector<Node *> orderedLeaves)
{
    if (orderedLeaves.size() == 1)
        return orderedLeaves.at(0);

    Node *newNode = new Node;
    newNode->left = orderedLeaves.at(0);
    newNode->right = orderedLeaves.at(1);
    newNode->symbols = orderedLeaves.at(0)->symbols + orderedLeaves.at(1)->symbols;
    newNode->weight = orderedLeaves.at(0)->weight + orderedLeaves.at(1)->weight;

    orderedLeaves.erase(orderedLeaves.begin(), orderedLeaves.begin() + 2);
    orderedLeaves.push_back(newNode);

    std::sort(orderedLeaves.begin(), orderedLeaves.end(),
              /* boost::bind actually*/
              NodeCmpFunc);

    return createTree(orderedLeaves);
}

string HuffmannTree::encodeHelper(Node *current, char letter)
{
    if (current->left == 0 && current->right == 0)
        return "";

    if (current->left->symbols.find(letter) != string::npos)
        return string("0") + encodeHelper(current->left, letter);
    if (current->right->symbols.find(letter) != string::npos)
        return string("1") + encodeHelper(current->right, letter);
    else
    {
        std::cerr << "No such symbol";
        throw std::exception();
    }
}

string HuffmannTree::decodeHelper(Node *current, string bits)
{
    if (bits.empty())
        return "";

    if (bits.at(0) == '1')
        current = current->right;
    else if (bits.at(0) == '0')
        current = current->left;

    if (current->left == 0 && current->right == 0)  // leaf
        return current->symbols + decodeHelper(root, string(bits.begin() + 1, bits.end()));
    else
        return decodeHelper(current, string(bits.begin() + 1, bits.end()));
}

HuffmannTree::HuffmannTree(const vector<SymbolInfo> &symbolsInfo)
{
    vector<SymbolInfo> symsInfo(symbolsInfo);

    std::sort(symsInfo.begin(), symsInfo.end());

    vector<Node *> leaves(symsInfo.size());
    for (size_t i = 0; i < symsInfo.size(); i++)
    {
        leaves[i] = new Node;
        leaves[i]->symbols = symsInfo[i].symbol;
        leaves[i]->weight = symsInfo[i].freq;
    }

    root = createTree(leaves);
}

string HuffmannTree::encode(string message)
{
    if (message.empty())
        return "";

    char letter = message.at(0);
    Node *current = root;
    return encodeHelper(current, letter) + encode(string(message.begin() + 1, message.end()));
}

string HuffmannTree::decode(string codedmsg)
{
    if (codedmsg.empty())
        return "";

    return decodeHelper(root, codedmsg);
}

File: ch2/HuffmanTree.py
# for the comparsion with C++ code

import operator

class Node:
    
    def __init__(self, symbols, weight):        
        self.left = None
        self.right = None
        self.symbols = symbols
        self.weight = weight


class HuffmanTree:
    
    def __createTree(self, leaves):
        if len(leaves) == 1:
            return leaves[0]
            
        newNode = Node(leaves[0].symbols + leaves[1].symbols,
                       leaves[0].weight + leaves[1].weight)
        newNode.left = leaves[0]
        newNode.right = leaves[1]

        leaves = leaves[2:]
        leaves.append(newNode)
        leaves.sort(key = lambda x: x.weight)
        return self.__createTree(leaves)

    def __init__(self, alphabet):
        sorted_alphabet = sorted(alphabet.items(), key = lambda x: x[1])
        
        leaves = []
        for k, v in sorted_alphabet:
            leaves.append(Node(k, v))

        self.root = self.__createTree(leaves)
        
    def encode(self, message):
        def helper(node, letter):
            if not node.left and not node.right:
                return ""
            
            if node.left.symbols.find(letter) != -1:
                return "0" + helper(node.left, letter)
            elif node.right.symbols.find(letter) != -1:
                return "1" + helper(node.right, letter)
            else:
                raise Exception
            
        if len(message) == 0:
            return ""
        else:
            return helper(self.root, message[0]) + self.encode(message[1:])
        
    def decode(self, codedMsg):
        def helper(node, bits):
            if len(bits) == 0:
                return ""
                
            if bits[0] == "0":
                node = node.left
            elif bits[0] == "1":
                node = node.right
            else:
                raise Exception
            
            if not node.left and not node.right:
                return node.symbols + helper(self.root, bits[1:])
            else:
                return helper(node, bits[1:])
            
        if len(codedMsg) == 0:
            return ""
            
        return helper(self.root, codedMsg)
    

if __name__ == "__main__":    
    print("Huffman tree")
    hf = HuffmanTree({ "A" : 5, "B" : 3, "C" : 16, "D" : 8, "E": 1 })
    
    msg = "ABBCCCCCCEDDDCAACCCCC"
    print("Message:", msg)
    
    encodedMsg = hf.encode(msg)
    print("Encoded message:", encodedMsg)
    
    decodedMsg = hf.decode(encodedMsg)
    print("Decoded message:", decodedMsg)

    print(decodedMsg == msg)
